stl/inc/deque

```diff
@@ -986,19 +986,68 @@ public:
     }
 
     void shrink_to_fit() {
-        size_type _Oldcapacity = _Block_size * _Mapsize();
-        size_type _Newcapacity = _Oldcapacity / 2;
+        if (empty()) {
+            if (_Map() != nullptr) {
+                _Reset_map();
+            }
+            return;
+        }
+
+        const auto _First_block_idx = static_cast<_Map_difference_type>(_Myoff() / _Block_size);
+        const auto _Last_block_idx =
+            static_cast<_Map_difference_type>(((_Myoff() + _Mysize() - 1) / _Block_size + 1) & (_Mapsize() - 1));
+
+        // deallocate unused blocks
 
-        if (_Newcapacity < _Block_size * _Minimum_map_size) {
-            _Newcapacity = _Block_size * _Minimum_map_size;
+        for (auto _Block_idx = _Map_difference_type{0}; _Block_idx != _First_block_idx; ++_Block_idx) {
+            auto& _Block_ptr = _Map()[static_cast<_Map_difference_type>(_Block_idx & (_Mapsize() - 1))];
+            if (_Block_ptr != nullptr) {
+                _Getal().deallocate(_Block_ptr, _Block_size);
+                _Block_ptr = nullptr;

```

Comment:
Some area in [0, x) ∈ [0, `_First_block_idx`) might still belong to active area...